### ISSUE TITLE
Upgrade to Solidity 0.5.13

### ISSUE THREAD

**Issue by travs** [melonproject/protocol#635]
Solidity 0.5 has some syntax changes that we may need to adapt to.
Also when we do this, we should change `Policy` back to an interface.

- [x] change version of non-third-party contracts
- [x] fix syntax of each of these where is doesn't work
- [x] change policy back to an interface
- [ ] fix `address payable` issue in `orderUpdateHook` (see comment)
- [ ] fix compiler warnings
- [x] Interfaces should only have external functions and should not be inherited by contracts (these should be enforced by the solidity 0.5.x compiler anyways)
- [x] Move all events from interfaces to their actual contracts
- [x] upgrade directly-depnded-on 3rd party contracts for which there is a new version (e.g. DSMath)

**Comment by travs** [comment]
From @SeanJCasey in #824 : we can also try to fix all warnings and update syntax with best practices (within reason)

There will still be a huge number of warnings because of the ABIEncoderV2, but other than that we can try to fix the warnings as well.

**Comment by SeanJCasey** [comment]
@travs @luongnt95 A couple things to add to the checklist above:

- Interfaces should only have `external` functions and should not be inherited by contracts (these should be enforced by the solidity 0.5.x compiler anyways)
- Move all events from interfaces to their actual contracts

**Comment by travs** [comment]
Just noting here that due to the strangeness of `address payable` being the type of `address(0)`, I had to change the one parameter of `orderUpdateHook` to `address payable[2] memory orderAddresses`. Note the `payable`. Couldn't yet figure out how to convert `address(0)` to an `address` and not `address payable`.

```solidity
// minimal example duplicating the compiler error
pragma solidity 0.5.13;

contract A {
    function inner(address[2] memory _input) public {}
    function outer() public {
        inner([address(0), address(0)]);
    }
}
```

```
solc A.sol

A.sol:6:15: Error: Invalid type for argument in function call.
Invalid implicit conversion from address payable[2] memory to address[2] memory requested.
inner([address(0), address(0)]);
```

Better `payable` type casting is coming [in 0.6.0](https://github.com/ethereum/solidity/pull/7349), so maybe there will be some improvements in the other direction as well

**Comment by travs** [comment]
Fixing this issue fixes #840 and fixes #841 (just cross-referencing)
